llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -2750,6 +2745,42 @@ static SDValue LowerIntrinsicVoid(SDValue Op, SelectionDAG &DAG) {
   return Op;
 }
 
+static SDValue lowerCTPOP(SDValue Op, SelectionDAG &DAG) {
+  SDValue V = Op->getOperand(0);
+  SDLoc DL(Op);
+
+  if (V.getValueType() == MVT::i16) {
+    SDValue Zext = DAG.getNode(ISD::ZERO_EXTEND, DL, MVT::i32, V);
+    SDValue CT = DAG.getNode(ISD::CTPOP, DL, MVT::i32, Zext);
+    return DAG.getNode(ISD::TRUNCATE, DL, MVT::i16, CT, SDNodeFlags::NoWrap);
+  }
+  if (V.getValueType() == MVT::i64) {
+    SDValue CT = DAG.getNode(ISD::CTPOP, DL, MVT::i32, V);
+    return DAG.getNode(ISD::ZERO_EXTEND, DL, MVT::i64, CT);
+  }
```

Comment:
This may be worth a comment that CTPOP node here is intentionally created with 64-bit input and 32-bit output and that we'll later pattern-match it to `popc` instruction.

Without the explanation this code looks as if we accidentally run CTPOP only only half of the bits of the input value.

Reply:
I ended up doing some more refactoring of this change, but I've added a comment to the new lowering function which hopefully explains what is going on.